src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java

```diff
@@ -577,6 +603,354 @@ public DataSet data()
+    {
+        private TxnTracesTable()
+        {
+            super(parse(VIRTUAL_ACCORD_DEBUG, TXN_TRACES,
+                        "Accord Transaction Traces",
+                        "CREATE TABLE %s (\n" +
+                        "  txn_id text,\n" +
+                        "  id_micros bigint,\n" +
+                        "  event_type text,\n" +
+                        "  at_micros bigint,\n" +
+                        "  command_store_id int,\n" +
+                        "  message text,\n" +
+                        "  PRIMARY KEY (txn_id, event_type, id_micros, at_micros)" +
+                        ')', UTF8Type.instance));
+        }
+
+        private AccordTracing tracing()
+        {
+            return ((AccordAgent)AccordService.instance().agent()).tracing();
+        }
+
+        @Override
+        protected void applyPartitionDeletion(ColumnValues partitionKey)
+        {
+            TxnId txnId = TxnId.parse(partitionKey.value(0));
+            tracing().eraseEvents(txnId);
+        }
+
+        @Override
+        protected void applyRangeTombstone(ColumnValues partitionKey, Range<ColumnValues> range)
+        {
+            TxnId txnId = TxnId.parse(partitionKey.value(0));
+            if (!range.hasLowerBound() || range.lowerBoundType() != BoundType.CLOSED) throw invalidRequest("May restrict deletion by at most one event_type");
+            if (range.lowerEndpoint().size() != 1) throw invalidRequest("Deletion restricted by lower bound on id_micros or at_micros is unsupported");
+            if (!range.hasUpperBound() || (range.upperBoundType() != BoundType.CLOSED && range.upperEndpoint().size() == 1)) throw invalidRequest("Range deletion must specify one event_type");
+            if (!range.upperEndpoint().value(0).equals(range.lowerEndpoint().value(0))) throw invalidRequest("May restrict deletion by at most one event_type");
+            if (range.upperEndpoint().size() > 2) throw invalidRequest("Deletion restricted by upper bound on at_micros is unsupported");
+            TraceEventType eventType = parseEventType(range.lowerEndpoint().value(0));
+            if (range.upperEndpoint().size() == 1)
+            {
+                tracing().eraseEvents(txnId, eventType);
+            }
+            else
+            {
+                long before = range.upperEndpoint().value(1);
+                tracing().eraseEventsBefore(txnId, eventType, before);
+            }
+        }
+
+        @Override
+        public void truncate()
+        {
+            tracing().eraseAllEvents();
+        }
+
+        @Override
+        public DataSet data()
+        {
+            SimpleDataSet dataSet = new SimpleDataSet(metadata());
+            tracing().forEach(id -> true, (txnId, eventType, permits, events) -> {
+                events.forEach(e -> {
+                    e.messages().forEach(m -> {
+                        dataSet.row(txnId.toString(), eventType.name(), e.atMicros, m.atNanos - e.atNanos)
```

Comment:
last column was supposed to be micros, not nanos (`id_micros, at_micros`)
is `atMicros` also on the right spot?